workers/price-reporter/src/manager.rs

```diff
@@ -97,29 +98,127 @@ impl AtomicPriceStreamState {
     }
 }
 
-/// A shareable mapping between exchange and the most recent state of their
-/// price stream for a given pair
+/// A shareable mapping between (exchange, base, quote) and the most recent
+/// state of the associated price stream
 #[derive(Clone, Debug)]
-pub struct ExchangeStates(HashMap<Exchange, Arc<AtomicPriceStreamState>>);
-
-impl ExchangeStates {
-    /// Construct a new exchange states instance from a set of exchanges
-    pub fn new_from_exchanges(exchanges: &[Exchange]) -> Self {
-        let mut exchange_states = ExchangeStates(HashMap::new());
-        for exchange in exchanges {
-            exchange_states.0.insert(*exchange, Arc::new(AtomicPriceStreamState::default()));
-        }
-        exchange_states
+pub struct SharedPriceStreamStates(
+    AsyncShared<HashMap<(Exchange, Token, Token), AtomicPriceStreamState>>,
```

Comment:
no real point in making the state atomic if it's wrapped in a lock, it might actually be net slower if atomic

Reply:
we only ever acquire a write lock on this object when adding an entry to it, when updating the atomic state it's just a read lock, so we don't add locking overhead to updating the atomic